File: ios/VocabReview/Services/VocabMergeService.swift
import Foundation

enum VocabMergeService {
    static func mergeReviewFields(local: VocabMap, into remote: VocabMap) -> VocabMap {
        var merged = remote

        for (word, localWord) in local {
            guard var remoteWord = merged[word] else {
                continue
            }

            remoteWord.status = localWord.status
            remoteWord.lastSeen = localWord.lastSeen
            remoteWord.reviewCount = localWord.reviewCount
            remoteWord.correctCount = localWord.correctCount
            remoteWord.missCount = localWord.missCount
            remoteWord.intervalDays = localWord.intervalDays
            remoteWord.dueAt = localWord.dueAt
            remoteWord.lastReviewed = localWord.lastReviewed

            merged[word] = remoteWord
        }

        return merged
    }
}

File: android/app/src/main/java/com/vocabreview/data/VocabMergeService.kt
package com.vocabreview.data

import com.vocabreview.model.VocabMap
import com.vocabreview.model.VocabWord

object VocabMergeService {
    /**
     * Merge local review changes into remote data.
     * Strategy: start with remote as base, only overwrite review fields for locally-dirty words.
     * Words deleted remotely are not re-added.
     */
    fun merge(local: VocabMap, remote: VocabMap, dirtyWords: Set<String>): VocabMap {
        val merged = remote.toMutableMap()

        for (word in dirtyWords) {
            val localEntry = local[word] ?: continue
            val remoteEntry = merged[word] ?: continue // skip if deleted remotely
            merged[word] = remoteEntry.copy(
                status = localEntry.status ?: remoteEntry.status,
                lastSeen = localEntry.lastSeen ?: remoteEntry.lastSeen,
                reviewCount = localEntry.reviewCount ?: remoteEntry.reviewCount,
                correctCount = localEntry.correctCount ?: remoteEntry.correctCount,
                missCount = localEntry.missCount ?: remoteEntry.missCount,
                intervalDays = localEntry.intervalDays ?: remoteEntry.intervalDays,
                dueAt = localEntry.dueAt ?: remoteEntry.dueAt,
                lastReviewed = localEntry.lastReviewed ?: remoteEntry.lastReviewed
            )
        }

        return merged
    }
}
